imports/wasi_snapshot_preview1/fs.go

```diff
@@ -905,7 +908,7 @@ const largestDirent = int64(math.MaxUint32 - wasip1.DirentSize)
 //
 // See https://github.com/WebAssembly/WASI/blob/snapshot-01/phases/snapshot/docs.md#fd_readdir
 // See https://github.com/WebAssembly/wasi-libc/blob/659ff414560721b1660a19685110e484a081c3d4/libc-bottom-half/cloudlibc/src/libc/dirent/readdir.c#L44
-func maxDirents(dir *sys.Readdir, bufLen uint32) (dirents []fsapi.Dirent, bufused, direntCount uint32, writeTruncatedEntry bool) {
+func maxDirents(dir fsapi.Readdir, bufLen uint32) (dirents []fsapi.Dirent, bufused, direntCount uint32, writeTruncatedEntry bool) {
```

Comment:
It appears that the only non-test usage of Peek is here. I haven't checked carefully, but I'm wondering if we can use Rewind in the case of not enough space vs Peek/advance always (plus needing to define and expose Peek)

Reply:
so my immediate reaction was "no" because if you are at the end of a window, Next() will discard that window and you will lose access to the last item; however, I may rework the window so that it actually slides and so that index 0 always contains the last item of the previous batch. index 1 would be the first item of the new batch. Extra care is needed when you are on batch #1 where index 0 would be meaningless.

Reply:
clever approach on making dirent -1 is always available internally for a rewind(pos-1)

I feel like when things move to a native approach, they'll use Seek and those impls won't need to hold state possibly, rather seek offset relative -1. That said, wishing things doesn't make them true ;)

Reply:
there is unfortunately another caveat, the way . and .. are now prepended is through another Readdir, so it's not as easy as moving the window... you have also to take into account the prepended Readdir...

Reply:
FWIW, back in the day I was thinking to make "bool addDotEntries" a parameter of the readDir function for this reason. Composition is difficult externally.

Reply:
I think I managed to make it work 😬